File: 백준/Silver/2644. 촌수계산/촌수계산.py
# BFS
from collections import deque
def BFS(graph, checker, start, end):
    queue = deque([(start, 0)]) # (노드명, 촌수)
    checker[start] = True

    while queue:
        start, cnt = queue.popleft() # (시작 노드명, 현재 촌수)
        if start == end:
            return cnt

        for next in graph[start]:
            if not checker[next]:
                queue.append((next, cnt+1)) # (다음 노드명, 현재 촌수+1)
                checker[next] = True
    
    return -1


# main
n = int(input())
start, end = map(int, input().split())
m = int(input())

graph = [[] for _ in range(n+1)]
for i in range(m):
    x, y = map(int, input().split())
    # 양방향 graph
    graph[x].append(y)
    graph[y].append(x)

checker = [False for _ in range(n+1)]

print(BFS(graph, checker, start, end))

File: 백준/Silver/2644. 촌수계산/촌수계산.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static String dfs_path = "";

    public static void main(String[] args) throws IOException {
        // 1.
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        st = new StringTokenizer(br.readLine());
        int start = Integer.parseInt(st.nextToken());
        int end = Integer.parseInt(st.nextToken());
        st = new StringTokenizer(br.readLine());
        int m = Integer.parseInt(st.nextToken());

        // 2. graph와 checker 생성
        ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
        for(int i=0; i<n+1; i++){
            graph.add(new ArrayList<Integer>());
        }
        for(int i=0; i<m; i++){
            st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            graph.get(x).add(y);
            graph.get(y).add(x);
        }
        boolean[] checker = new boolean[n+1];
        
        // 3. 촌수 계산
        System.out.println(BFS(graph, checker, start, end));
    }

    public static int BFS(ArrayList<ArrayList<Integer>> graph, boolean[] checker, int start, int end){
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{start, 0}); // [노드 번호, 탐색하는 너비의 순서]
        checker[start] = true;

        while(!queue.isEmpty()){
            int[] temp = queue.poll();
            int init = temp[0]; // 부모 노드 번호
            int cnt = temp[1]; // 부모 노드가 속한 너비의 순서

            if(init == end){
                return cnt;
            }

            for(int next: graph.get(init)){
                if (!checker[next]){
                    queue.add(new int[]{next, cnt+1}); // [자식 노드 번호, 자식 노드가 속한 너비의 순서(부모+1)]
                    checker[next] = true;
                }
            }
        }

        return -1;
    }
}